docs: use tree id so the current time is used as mod time of each file

fastnloud opened the pull request:
So I've been willing to try FrankenPHP for a while now to create a static binary. After following the docs I got it up and running quite easily.

But, after a while (say 2 days), I was getting "file not found" errors when running the `php-cli` and notices _some_ files in the the local temp. directory where missing. Compiling the binary again fixes the problem. Also, removing the directory altogether fixes the problem since then it will recreate _all_ files.

After some digging I found out that `git archive HEAD` uses the commit time as modification time of each file. Could this cause the cache to be partially deleted? (older files first)

- https://git-scm.com/docs/git-archive#_description

Note that I cannot (yet) confirm that this actually fixes the "file not found" errors I'm getting. I will see what it does after 2 or 3 days.

At any case, is this a know issue? 🤔

dunglas commented:
On startup, the PHP app contained in the binary is extracted in the system tmp directory.
Maybe have you a cron jobs or something like that removing temporary files after some time? That would explain the issue.

fastnloud commented:
@dunglas not that I'm aware of. I kind of assumed that these temp. files get cleaned by system processes. I'm using macOS 14.x by the way.

I found some related questions on that topic:

- https://superuser.com/a/187105
- https://apple.stackexchange.com/a/404401

Running `man confstr`:

```
...
DESCRIPTION
     ...
     _CS_DARWIN_USER_TEMP_DIR
             Provides the path to a user's temporary items directory. The directory will be created it if does not already exist. This directory is created with access permissions of 0700 and
             restricted by the umask(2) of the calling process and is a good location for temporary files.

             By default, files in this location may be cleaned (removed) by the system if they are not accessed in 3 days.
...
```

It look like there is _some_ process then that gets rid of these files. 

I'm wondering, if this is the case, then perhaps these files should use the mod time on startup to avoid these files being deleted by this (or any other) process?

Like to hear what you think.

dunglas commented:
Or maybe can we store them in a non/temporary directory to prevent this issue?

fastnloud commented:
> Or maybe can we store them in a non/temporary directory to prevent this issue?

In this case I think storing these files in a temp. directory is good because whenever the binary gets updated or deleted these unused files then get deleted automatically.

We have no uninstaller to do so or you will have to do it manually.

withinboredom commented:
So, after a bit of research, it looks like the only way to prevent cleanup is to keep all the files open with a file handle. Naturally, a reboot would put the entire temporary directory in an unknown state. It will delete whatever it wants, for whatever reason.

This is the same on all *nix/osx.

So, if Frankenphp is using a stable directory, it needs to make sure that every startup results in verifying the folder integrity and fixing any missing/changed folders. Then it needs to open every single file in the directory to prevent the OS from purging the files.

The first part is feasible and realistic, opening every file is probably not practical. There's some evidence that access time may play a role in osx purging, but nothing about it in any official sense.

I will go digging through C#'s standalone mode and see if there are any tricks there (it uses this same method).

fastnloud commented:
> Note that I cannot (yet) confirm that this actually fixes the "file not found" errors I'm getting. I will see what it does after 2 or 3 days.

Well, my "fix" did not fix the "file not found" errors I'm getting. I was hoping the temp. directory (`frankenphp_b79d1654713dbd363e30a9a0c95a6557`) would get deleted completely with the mod time being the same for all files.

This is unfortunately not the case:

```
Warning: require(/private/var/folders/60/nqp4rch947l8y_8l3dvwkhjc0000gn/T/frankenphp_b79d1654713dbd363e30a9a0c95a6557/vendor/composer/ClassLoader.php): Failed to open stream: No such file or directory in /private/var/folders/60/nqp4rch947l8y_8l3dvwkhjc0000gn/T/frankenphp_b79d1654713dbd363e30a9a0c95a6557/vendor/composer/autoload_real.php on line 12

Fatal error: Uncaught Error: Failed opening required '/private/var/folders/60/nqp4rch947l8y_8l3dvwkhjc0000gn/T/frankenphp_b79d1654713dbd363e30a9a0c95a6557/vendor/composer/ClassLoader.php' (include_path='.:') in /private/var/folders/60/nqp4rch947l8y_8l3dvwkhjc0000gn/T/frankenphp_b79d1654713dbd363e30a9a0c95a6557/vendor/composer/autoload_real.php:12
Stack trace:
#0 /private/var/folders/60/nqp4rch947l8y_8l3dvwkhjc0000gn/T/frankenphp_b79d1654713dbd363e30a9a0c95a6557/vendor/composer/autoload_real.php(28): ComposerAutoloaderInita043caa845dd0c427a6933293ced6c32::loadClassLoader('Composer\\Autolo...')
#1 /private/var/folders/60/nqp4rch947l8y_8l3dvwkhjc0000gn/T/frankenphp_b79d1654713dbd363e30a9a0c95a6557/vendor/autoload.php(25): ComposerAutoloaderInita043caa845dd0c427a6933293ced6c32::getLoader()
#2 /private/var/folders/60/nqp4rch947l8y_8l3dvwkhjc0000gn/T/frankenphp_b79d1654713dbd363e30a9a0c95a6557/bin/platform-extension(8): require('/private/var/fo...')
#3 {main}
  thrown in /private/var/folders/60/nqp4rch947l8y_8l3dvwkhjc0000gn/T/frankenphp_b79d1654713dbd363e30a9a0c95a6557/vendor/composer/autoload_real.php on line 12
```

Probably because the mod times did not get applied to vendor files.

dunglas commented:
IMHO the safest fix would be to uncompress the app in outside of `/tmp` and purge old directories when a new one version of the embedded app is available.

fastnloud commented:
> IMHO the safest fix would be to uncompress the app in outside of `/tmp` and purge old directories when a new one version of the embedded app is available.

You're probably right. I just wonder how the app knows what old directory to purge when you are using another version of the app. Also, how can we purge any files when you decide te remove the app? Or is this something you'd have to do manually then?

Meanwhile I'm running another test using `touch -t` on every file after installation and before running the `./build-static.sh` script:

```zsh
find $TMPDIR/my-prepared-app -print | while read filename; do
    touch -t $now $filename
done
```

@withinboredom what is your take on this? 🤔

withinboredom commented:
I'm curious, are your files getting deleted during a reboot cycle or while running?

So as far as I can tell, C# keeps all files open for the duration of the execution. In *nix's, this prevents the files from being deleted. However, they also delete the directory on application shutdown and use a unique folder per execution (to prevent reboots from messing up the directory).

It may be interesting to continue with the current logic but allow specifying an `--install` directory which points to a location that it will be extracted to and executed. An uninstaller is usually a separate program/process, but it should be sufficient just to delete the provided folder.

I personally think that would be the best solution, a temporary directory for one-off runs or development/testing; and a proper `--install` directory for production deployments.

fastnloud commented:
@withinboredom looks like this happens when my computer is asleep. So technically still running(?), but not on a reboot cycle at least.

I can elaborate a bit further:

- I'm not running the app in the background
- I'm using the app locally
- I'm **only** using the app to run `php-cli` since it's a CLI tool

I had a look at the code and noticed on init it will only check if the embedded app temp. directory exists. Why not always untar and check the files individually? If necessary, I'm willing to make a new pull request since then the docs does not require any changes.

withinboredom commented:
I guess it depends on the level of sleep. If it actually is storing RAM onto disk and shutting down, then likely when the kernel reboots it will clean /tmp before/during restoring from RAM. Checking file integrity will probably fix it.

withinboredom commented on embed.go:
```diff
@@ -97,37 +95,39 @@ func untar(dir string) (err error) {
-					// for now.
-					log.Printf("error changing modtime: %v (further Chtimes errors suppressed)", err)
-					loggedChtimesError = true // once is enough
+			if _, err := os.Stat(abs); os.IsNotExist(err) {
```
Do we also want to check if the file is different than what we expect by looking at the mtime?

dunglas replied:
Not sure if it's a good idea: if the user modifies in a dirty way a file uncompressed in the FS, we may want to keep it (new versions of the app will create a different folder anyway).

withinboredom reviewed (approved):
I think this would work well. LGTM. @dunglas feel free to merge if you agree.

fastnloud commented:
> I think this would work well. LGTM. @dunglas feel free to merge if you agree.

Thanks @withinboredom 👍 I rebased and dropped the commit on the docs as this does not fix anything so it makes no sense to merge.

dunglas commented:
Thank you very much @fastnloud!
